### PR TITLE
logging/anomaly: Support configuration filter types

### PR DESCRIPTION
**Pull request by jlucovsky**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [2958](https://redmine.openinfosecfoundation.org/issues/2958)

* Continuation of #4033  
* Requires [Suricata-verify PR #109](https://github.com/OISF/suricata-verify/pull/109) 

Updates since last PR:
- Reworked configuration settings controlling anomaly logging (see discussion in #4033)

### CODE REVIEW

**Comment by jlucovsky** [comment]
Re-opened; depending on s-v pr #109.

File: src/output-json-anomaly.c [hunk 1]
```diff
@@ -348,9 +361,29 @@ static void SetFlag(const ConfNode *conf, const char *name, uint16_t flag, uint1
...
+        /* Check for metadata to enable/disable. */
+        ConfNode *typeconf = ConfNodeLookupChild(conf, "types");
+        if (typeconf != NULL) {
+            SetFlag(typeconf, "application-layer", LOG_JSON_APPPROTO_TYPE, &flags);
+            SetFlag(typeconf, "packet", LOG_JSON_PACKET_TYPE, &flags);
+        }
+        SetFlag(conf, "packethdr",   LOG_JSON_PACKETHDR,   &flags);
+    }
+    if (((flags & (LOG_JSON_PACKET_TYPE | LOG_JSON_PACKETHDR)) == LOG_JSON_PACKETHDR) && !_warn_no_packet) {
+        SCLogNotice("Anomaly logging configured to include packet headers, however packet "
```
Comment (victorjulien) [inline src/output-json-anomaly.c]: I think this should be a SCLogWarning instead.

File: src/output-json-anomaly.c [hunk 2]
```diff
@@ -348,9 +361,29 @@ static void SetFlag(const ConfNode *conf, const char *name, uint16_t flag, uint1
...
+                    "type logging has not been selected. Packet headers will not be logged.");
+        _warn_no_packet = true;
+        flags &= ~LOG_JSON_PACKETHDR;
+    }
+
+    if (flags == 0 && !_warn_no_flags) {
+        SCLogNotice("Anomaly logging has been configured; however, no logging types "
```
Comment (victorjulien) [inline src/output-json-anomaly.c]: SCLogWarning

File: doc/userguide/output/eve/eve-json-output.rst [hunk 3]
```diff
@@ -86,19 +86,38 @@ Anomalies are event records created when packets with unexpected or anomalous
...
             # occurrences of anomalies may experience packet processing degradation.
-
-            # Enable dumping of packet header
-            # packethdr: no            # enable dumping of packet header
+            #
+            # Anomalies are reported for the following:
+            # 1. Packet: Values and conditions that are detected while decoding individual packets. This includes
+            # invalid or unexpected values for low-level protocol lengths as well as stream related events (TCP 3-way
+            # handshake issues, unexpected sequence number, etc).
+            # 2. Application level: These denote application layer specific conditions that are unexpected, invalid or
```
Comment (victorjulien) [inline doc/userguide/output/eve/eve-json-output.rst]: this line is slightly too long as is showing by the 'or' wrapping around

File: suricata.yaml.in [hunk 4]
```diff
@@ -160,9 +160,23 @@ outputs:
...
-
-            # Enable dumping of packet header
-            # packethdr: no            # enable dumping of packet header
+            #
+            # Anomalies are reported for the following:
+            # 1. Packet: Values and conditions that are detected while decoding individual packets. This includes
+            # invalid or unexpected values for low-level protocol lengths as well as stream related events (TCP 3-way
+            # handshake issues, unexpected sequence number, etc).
+            # 2. Application level: These denote application layer specific conditions that are unexpected, invalid or
+            # are unexpected given the application monitoring state.
```
Comment (victorjulien) [inline suricata.yaml.in]: Can we distinguish between decoder and stream events too?

File: doc/userguide/output/eve/eve-json-output.rst (hunk 3 above)
Comment (jlucovsky) [reply doc/userguide/output/eve/eve-json-output.rst]: Will correct.

File: src/output-json-anomaly.c (hunk 1 above)
Comment (jlucovsky) [reply src/output-json-anomaly.c]: Fixed.

File: suricata.yaml.in (hunk 4 above)
Comment (jlucovsky) [reply suricata.yaml.in]: Of course. Let me know if you want that included with this work.

Comment (victorjulien) [reply suricata.yaml.in]: What works best for you. I would like to get the stream events logic in soon so it can go in to 5.0rc1

Comment (victorjulien) [reply suricata.yaml.in]: Yeah, I think it is likely that ppl will want to disable stream events separately from the decoder events.

Comment (jlucovsky) [reply suricata.yaml.in]: > Should this be a one-to-one mapping with the anomaly.type field in the logged output? So we'd have:
> 
> * packet
> * applayer
>   etc...
> 
> This is how eve logging is done where event_type matches the types list.

I'll ensure that the "type" matches one of decode, stream, application-layer in the reworked pr.

**Comment by jlucovsky** [comment]
Continued in #4142

**Pull request closed without merging**